js/components/FileChooser.tsx

```diff
@@ -0,0 +1,215 @@
+                filesRef.current = newFiles;
+                onChange(newFiles);
+            }
+        }
+    }, []);
+
+    const extraProps = useMemo<ExtraProps>(() => {
+        // Accept all of the given MIME types and extensions. An argument
+        // could be made to throw out all of the MIME types and just keep
+        // the extensions, but this seemed to follow the intent in the
+        // spec:
+        //
+        // https://html.spec.whatwg.org/multipage/input.html#file-upload-state-(type=file)
+        //
+        // which allows pretty generous interpretations.
+        //
+        const newAccept = [];
+        for (const type of accept) {
+            for (const [typeString, suffixes] of Object.entries(type.accept)) {
+                newAccept.push(typeString);
+                if (Array.isArray(suffixes)) {
+                    newAccept.push(...suffixes);
+                } else {
+                    newAccept.push(suffixes);
+                }
+            }
+        }
+
+        const extraProps: { accept?: string } = {};
+        if (newAccept.length > 0) {
+            extraProps['accept'] = newAccept.join(',');
+        }
+
+        return extraProps;
+    }, [accept]);
+
+    return (
+        <input type="file" role='button' aria-label='Open file'
+            ref={inputRef}
+            onChange={onChangeInternal}
+            disabled={disabled}
+            {...extraProps} />
+    );
+};
+
+interface FilePickerChooserProps {
+    disabled?: boolean;
+    onChange?: (files: FileSystemFileHandle[]) => void;
+    accept?: FilePickerAcceptType[];
+}
+
+const FilePickerChooser = ({
+    disabled = false,
+    onChange = () => { },
+    accept = [ACCEPT_EVERYTHING_TYPE]
+}: FilePickerChooserProps) => {
+    const [busy, setBusy] = useState<boolean>(false);
+    const filenameSpanRef = useRef<HTMLSpanElement>(null);
+    const fileHandlesRef = useRef<FileSystemFileHandle[]>();
+
+    const onClickInternal = useCallback(async () => {
+        if (busy) {
+            return;
+        }
+        setBusy(true);
+        try {
+            const pickedFiles = await window.showOpenFilePicker({
+                multiple: false,
+                excludeAcceptAllOption: true,
+                types: accept,
+            });
+            if (fileHandlesRef.current !== pickedFiles) {
+                fileHandlesRef.current = pickedFiles;
+                onChange(pickedFiles);
+            }
+        } catch (e: unknown) {
+            console.error(e);
+        } finally {
+            setBusy(false);
+        }
+    }, []);
+
+    useEffect(() => {
+        if (!filenameSpanRef.current) {
+            return;
+        }
+
+        if (fileHandlesRef.current?.length) {
+            filenameSpanRef.current.textContent = fileHandlesRef.current[0].name;
+        } else {
+            filenameSpanRef.current.textContent = 'No file selected';
+        }
+    }, [fileHandlesRef.current]);
+
```

Comment:
I would avoid using a ref to manipulate the DOM here, you can do something like:

```tsx
     const [selectedFilename, setSelectedFilename] = useState<string>()
```
```tsx
    setSelectedFilename(
        fileHandles.current?.length ?
        fileHandles.current?[0].name :
        'No file selected';
    )
```
```tsx
         <span role="label">{selectedFilename}</span>
```